Feature/frontend upload (#96)
madewithkoji/docs: src/asciidoc/docs/guidelines/frontendUploading.adoc

To upload files with `generateSignedUploadRequest`, you configure the backend to acquire a presigned URL, which grants temporary access for uploading a file.
Then, you configure the frontend to retrieve a presigned URL from the backend and use it to upload a file.

== Prerequisites

This guide covers advanced topics.
It is suggested that you are familiar with the following topics before proceeding.

* React and ES6 basics.
* The Koji editor, <<instant-remixing#, remix process>>, and the <<withkoji-vcc-package,@withkoji/vcc>> package.
* https://expressjs.com/[Express.js] web framework.
* Backend communication and the <<withkoji-database-package,@withkoji/database-sdk>> package.
* Performing HTTP requests (this example uses https://github.com/axios/axios[axios]).

== Configuring the backend

To upload from the frontend, you first need to acquire a presigned URL by using the Koji database SDK.

In a backend service for your Koji project, set up an Express.js server with a POST method to the `/get-upload-url` endpoint.
From this endpoint, call the following function.

[source, javascript]
----
import Database from '@withkoji/database';

> const uploadImage = async (req, res) => {

Reviewer: Might be good to change this function to be called `getUploadURL` or something similar?

  try {
    const { filename } = req.body;

    const db = new Database({
      projectId: res.locals.KOJI_PROJECT_ID,
      projectToken: res.locals.KOJI_PROJECT_TOKEN,
    });

    const data = await db.generateSignedUploadRequest(filename);

    res.status(200).json({ data });
  } catch (err) {
    console.log('err', err);
    res.status(500).json({ err: err.toString() });
  }
};

export default uploadImage;
----

== Configuring the frontend

From your frontend service, call the `/get-upload-url` endpoint to retrieve the presigned URL.

[source,javascript]
----

> const { data: { data: { signedRequest, url } } = {} } = await axios({

Reviewer: Might be good to show the structure of the data that is returned along with a brief explanation. While it's okay that this thing feels like magic, it would be helpful for developers to know that this is really just a wrapper around an S3 method and if they see that it's similar to what they've used before may grok it more quickly.

  method: 'post',
  url: `${serviceMap.backend}/get-upload-url`,
  data: {
    filename: file.name,
  },
  headers: {
    'Content-Type': 'application/json',
  },
});
----

Create form data and include the presigned URL data and the file object to be uploaded.

[source,javascript]
----
// Build form data to send to s3
const formData = new FormData();

// Attach each of the request fields to the form
Object.keys(signedRequest.fields).forEach((key) => {
  formData.append(key, signedRequest.fields[key]);
});

// Append file last
formData.append('file', file);
----

> Post this data to the URL, and then check for a successful status (since there is no response data).

Reviewer: It could be helpful to show even a basic use case of what this would look like. Something like 

```
const image = new window.Image()
image.onload = () => {
  document.body.append(image);
}
image.src = signedRequest.url;
```

* https://expressjs.com/[Express.js] web framework.
* Backend communication and the <<withkoji-database-package,@withkoji/database-sdk>> package.
* Performing HTTP requests (this example uses https://github.com/axios/axios[axios]).

== Configuring the backend

To upload from the frontend, you first need to acquire a presigned URL by using the Koji database SDK.

In a backend service for your Koji project, set up an Express.js server with a POST method to the `/get-upload-url` endpoint.
From this endpoint, call the following function.

[source, javascript]
----
import Database from '@withkoji/database';

const getUploadURL = async (req, res) => {
  try {
    const { filename } = req.body;

    const db = new Database({
      projectId: res.locals.KOJI_PROJECT_ID,
      projectToken: res.locals.KOJI_PROJECT_TOKEN,
    });

    const data = await db.generateSignedUploadRequest(filename);

    res.status(200).json({ data });
  } catch (err) {
    console.log('err', err);
    res.status(500).json({ err: err.toString() });
  }
};

export default getUploadURL;
----

== Configuring the frontend

From your frontend service, call the `/get-upload-url` endpoint to retrieve the presigned URL.

> For more information on the response value see <<#_signed_request_response>>.

Reviewer 2: ```suggestion
For more information on the response value, see <<#_signed_request_response>>.
```

  method: 'post',
  url: `${serviceMap.backend}/get-upload-url`,
  data: {
    filename: file.name,
  },
  headers: {
    'Content-Type': 'application/json',
  },
});
----

Create form data and include the presigned URL data and the file object to be uploaded.

[source,javascript]
----
// Build form data to send to s3
const formData = new FormData();

// Attach each of the request fields to the form
Object.keys(signedRequest.fields).forEach((key) => {
  formData.append(key, signedRequest.fields[key]);
});

// Append file last
formData.append('file', file);
----

Post this data to the URL, and then check for a successful status (since there is no response data).
If the status is `204`, you can now use `signedRequest.url` which will point to your uploaded file.

[source,javascript]
----
// Post to s3
const { status } = await axios({
  headers: { 'content-type': 'application/x-www-form-urlencoded' },
  method: 'POST',
  data: formData,
  url: signedRequest.url,
});

if (status === 204) {
  const image = new window.Image()
  image.onload = () => {
    document.body.append(image);
  }
  image.src = signedRequest.url;
}
----

== Signed request response

> As the `generateSignedUploadRequest` method is a wrapper for the https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-authentication-HTTPPOST.html[presigned upload request in Amazon S3] the object returned from the `get-upload-url` endpoint is equivalent, all the data stored within the `signedRequest` property are related to the Amazon Web Services functionaliy.

Reviewer 2: ```suggestion
As the `generateSignedUploadRequest` method is a wrapper for the https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-authentication-HTTPPOST.html[presigned upload request in Amazon S3], the return objects are equivalent.
The data in the `signedRequest` property relates to the Amazon Web Services functionality.
```

Fixed a typo and tried to make this a little shorter. Might not be the best rewriting so feel free to edit again.

The response object includes the follow properties:

[source, javascript]
----
{
  data: {
    signedRequest: {
      url: string; <1>
      fields: {
        Policy: string; <2>
        X-Amz-Algorithm: string; <3>
        X-Amz-Credential: string; <4>
        X-Amz-Date: string; <5>
        X-Amz-Signature: string; <6>
        bucket: string; <7>
        key: string; <8>
      }
    },
    url: string; <9>
  }
}
----

> <1> `data.signedRequest.url` – The upload URL which points to the S3 bucket and where the file be sent.

Reviewer 2: ```suggestion
<1> `data.signedRequest.url` – The upload URL that points to the S3 bucket and where the file be sent.
```

> <2> `data.signedRequest.fields.Policy` – The Base64-encoded security policy for the request .

Reviewer 2: ```suggestion
<2> `data.signedRequest.fields.Policy` – The Base64-encoded security policy for the request.
```

> <3> `data.signedRequest.fields.X-Amz-Algorithm` – The signing algoruthm used.

Reviewer 2: ```suggestion
<3> `data.signedRequest.fields.X-Amz-Algorithm` – The signing algorithm used.
```

<4> `data.signedRequest.fields.X-Amz-Credential` – The Amazon scope credentials used to sign this URL.

> <5> `data.signedRequest.fields.X-Amz-Date` – The ISO8601 formatted date of when the URL was signed.

Reviewer 2: ```suggestion
<5> `data.signedRequest.fields.X-Amz-Date` – The ISO8601-formatted date when the URL was signed.
```

<6> `data.signedRequest.fields.X-Amz-Signature` – The HMAC-SHA256 hash of the Amazon security policy.
<7> `data.signedRequest.fields.bucket` – The Amazon S3 bucket name.
<8> `data.signedRequest.fields.key` – The Amazon S3 object key.

> <9> `data.url` – The destination URL, the file will be accessible via this URL once uploaded.

Reviewer 2: ```suggestion
<9> `data.url` – The destination URL.
The file will be accessible via this URL when uploaded.
```
